### DISCUSSION TITLE
Is there a way to hook into the current code parser?

### DISCUSSION THREAD

**Discussion by snake-py** [microsoft/vscode-discussions#164] [Extension Development QnA] [answered]
I want to write a documentation extension and it would be great if I can somehow infer the current peace of code, e.g. if the user works on a function or on a class, I want to identify that it is a function or a class a look if it is used in the rest of the project. 

Will I need to provide a file parser for every language myself or can I hook into the parser which is currently used for that file?

**Comment by r3m0t** [comment, marked as answer]
Have a look at these commands.

https://code.visualstudio.com/api/references/commands

If you query document symbols (which show up in the Outline pane) and compare to cursor position you'll know what function they were working on.

You could use the references provider (which provides Shift+F12) to check if the function is used in other locations.

Results will vary quite a bit by language though, ie depending on the LSP used for that language you could use a lot of CPU or memory or make the IDE less responsive for when the user actually wants to use an LSP feature like F12.

You might want to look at using a fast parser like tree-sitter so that your functionality is less reliant on the exact extensions the user has installed. As Typescript and Javascript are the only languages with support built into VS Code.
